llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -312,22 +314,161 @@ class RegionEndEdges : public ScheduleDAGMutation {
+/// dependencies related to EntrySU in other mutations
 class EmitFixedSUnits : public ScheduleDAGMutation {
 public:
+  // The purpose of this class is to implement a mechanism that tracks the
+  // cycles in which each register is last written or read. This information
+  // allows us to identify safe distances—measured in cycle units—where
+  // instructions can be placed without conflicts. These safe distances take
+  // into account read-after-write (RAW), write-after-read (WAR), and
+  // write-after-write (WAW) dependencies among all unscheduled instructions and
+  // timed bundles. This includes both top-insert instructions and their
+  // predecessor instructions in pipelined loops.
+  struct RegisterEventTracker {
+    using fixed_iterator = MachineBasicBlock::iterator;
+    const InstrItineraryData *InstrItins;
+    const TargetRegisterInfo *TRI;
+    std::map<MCRegister, unsigned> RegisterToCycleDef;
+    std::map<MCRegister, unsigned> RegisterToCycleUse;
+
+    RegisterEventTracker(const InstrItineraryData *Itins,
+                         const TargetRegisterInfo *TRI)
+        : InstrItins(Itins), TRI(TRI) {};
+
+    const std::map<MCRegister, unsigned> &getRegToCycleMap(bool IsDef) const {
+      return IsDef ? RegisterToCycleDef : RegisterToCycleUse;
+    }
+
+    std::map<MCRegister, unsigned> &getRegToCycleMap(bool IsDef) {
+      return const_cast<std::map<MCRegister, unsigned> &>(
+          const_cast<const RegisterEventTracker *>(this)->getRegToCycleMap(
+              IsDef));
+    }
+
+    void updateUseDefMaxCycle(Register Reg, unsigned Latency, bool IsDef) {
+      std::map<MCRegister, unsigned> &RegisterToCycle = getRegToCycleMap(IsDef);
+      MCRegister MCReg = Reg.asMCReg();
+      auto RegCycle = RegisterToCycle.find(MCReg);
+      if (RegCycle == RegisterToCycle.end())
+        RegisterToCycle[MCReg] = Latency;
+      else
+        RegisterToCycle[MCReg] = std::max(RegCycle->second, Latency);
+    }
+
+    // This function calculates the cycles in which each register will be last
+    // materialized or used, taking into account a specified sequence of timed
+    // bundles. PastTheEndCycles is utilized to project the availability cycle
+    // to the subsequent region.
+    void computeUseDefForward(ArrayRef<AIE::MachineBundle> Bundles,
+                              bool PastTheEndCycles) {
+      int Cycle = 0;
+      const int TotalCycles = Bundles.size();
+      for (const auto &Bundle : Bundles) {
+        for (MachineInstr *BundledMI : Bundle.getInstrs()) {
+          const unsigned SchedClass = BundledMI->getDesc().getSchedClass();
+          for (unsigned OpNum = 0; OpNum < BundledMI->getNumOperands();
+               OpNum++) {
+            const MachineOperand &MO = BundledMI->getOperand(OpNum);
+            if (!MO.isReg())
+              continue;
+            const bool IsDef = MO.isDef();
+            // Use past the end is not relevant.
```

Comment:
Why is that? Don't we want to gather all use/def events so that we can e.g. connect a def in the region to a use "past the end"?